libs/execution_context/ebpf_native.c

```diff
@@ -0,0 +1,1219 @@
+        _ebpf_native_client_attach_callback,
+        _ebpf_native_client_detach_callback);
+
+    if (return_value != EBPF_SUCCESS) {
+        goto Done;
+    }
+
+Done:
+    if (return_value != EBPF_SUCCESS) {
+        if (hash_table_created) {
+            ebpf_hash_table_destroy(_ebpf_native_client_table);
+        }
+        ebpf_lock_destroy(&_ebpf_native_client_table_lock);
+    }
+
+    return return_value;
+}
+
+static ebpf_native_map_t*
+_ebpf_native_get_next_map_to_create(_In_ ebpf_native_map_t* maps, size_t map_count)
+{
+    for (uint32_t i = 0; i < map_count; i++) {
+        ebpf_native_map_t* map = &maps[i];
+        if (map->handle != ebpf_handle_invalid) {
+            // Already created.
+            continue;
+        }
+        if (map->entry->definition.type != BPF_MAP_TYPE_ARRAY_OF_MAPS &&
+            map->entry->definition.type != BPF_MAP_TYPE_HASH_OF_MAPS) {
+            return map;
+        }
+        if (map->inner_map == NULL) {
+            // This map requires an inner map template, look up which one.
+            for (uint32_t j = 0; j < map_count; j++) {
+                ebpf_native_map_t* inner_map = &maps[i];
+                if (!inner_map) {
+                    continue;
+                }
+                if (inner_map->original_fd == map->inner_map_original_fd) {
+                    map->inner_map = inner_map;
+                    break;
+                }
+            }
+            if (map->inner_map == NULL) {
+                // We can't create this map because there is no inner template.
+                continue;
+            }
+        }
+        if (map->inner_map->handle == ebpf_handle_invalid) {
+            // We need to create the inner map template first.
+            continue;
+        }
+
+        return map;
+    }
+
+    // There are no maps left that we can create.
+    return NULL;
+}
+
+static ebpf_result_t
+_ebpf_native_initialize_maps(_In_ ebpf_native_map_t* native_maps, _In_ map_entry_t* maps, size_t map_count)
+{
+    ebpf_result_t result = EBPF_SUCCESS;
+    const int ORIGINAL_FD_OFFSET = 1;
+    for (uint32_t i = 0; i < map_count; i++) {
+        if (maps[i].definition.pinning != PIN_NONE && maps[i].definition.pinning != PIN_GLOBAL_NS) {
+            result = EBPF_INVALID_ARGUMENT;
+            goto Done;
+        }
+        native_maps[i].entry = &maps[i];
+        native_maps[i].original_fd = i + ORIGINAL_FD_OFFSET;
+        native_maps[i].handle = ebpf_handle_invalid;
+        maps[i].address = NULL;
+
+        if (maps[i].definition.pinning == PIN_GLOBAL_NS) {
+            // Construct the pin path.
+            size_t prefix_length = strnlen(DEFAULT_PIN_ROOT_PATH, EBPF_MAX_PIN_PATH_LENGTH);
+            size_t name_length = strnlen_s(maps[i].name, BPF_OBJ_NAME_LEN);
+            if (name_length == 0 || name_length >= BPF_OBJ_NAME_LEN ||
+                prefix_length + name_length + 1 >= EBPF_MAX_PIN_PATH_LENGTH) {
+                result = EBPF_INVALID_ARGUMENT;
```

Comment:
Add logging for all cases where the root cause of the failure might be user error.

Reply:
Added.